gadget/gadget.go

```diff
@@ -206,6 +208,107 @@ type VolumeUpdate struct {
+// to always find the device again, so we need to do a search, trying to find a
+// device which matches each trait in turn, and verify it matches the  physical
+// structure layout and if not move on to using the next trait.
+type DiskVolumeDeviceTraits struct {
+	// each member here is presented in descending order of certainty about the
+	// likelihood of being compatible if a candidate physical device matches the
+	// member. I.e. CachedDevicePath is more trusted than CachedKernelPath is
+	// more trusted than DiskID is more trusted than using the MappedStructures
+
+	// CachedDevicePath is the device path in sysfs and in /dev/disk/by-path the
+	// volume was measured and observed at during UC20+ install mode.
+	CachedDevicePath string `json:"cached-device-path"`
+
+	// CachedKernelPath is the device path like /dev/vda the volume was
+	// measured and observed at during UC20+ install mode.
+	CachedKernelPath string `json:"cached-kernel-path"`
+
+	// DiskID is the disk's identifier, it is a UUID for GPT disks or an
+	// unsigned integer for DOS disks encoded as a string in hexadecimal as in
+	// "0x1212e868".
+	DiskID string `json:"disk-id"`
+
+	// Structure contains trait information about each individual structure in
+	// the volume that may be useful in identifying whethere a disk matches a
+	// volume or not.
+	Structure []DiskStructureDeviceTraits `json:"mapped-structures"`
+}
+
+// DiskStructureDeviceTraits is a similar to DiskVolumeDeviceTraits, but is a
+// set of traits for a specific structure on a disk rather than the full disk
+// itself.
+type DiskStructureDeviceTraits struct {
+	// CachedDevicePath is the device path in sysfs and in /dev/disk/by-path the
+	// partition was measured and observed at during UC20+ install mode.
+	CachedDevicePath string `json:"cached-device-path"`
+	// CachedKernelPath is the device path like /dev/vda1 the partition was
+	// measured and observed at during UC20+ install mode.
+	CachedKernelPath string `json:"cached-kernel-path"`
+	// PartitionUUID is the partuuid as defined by i.e. /dev/disk/by-partuuid
+	PartitionUUID string `json:"partition-uuid"`
+	// FilesystemLabel is the label of the filesystem for structures that have
+	// filesystems, i.e. /dev/disk/by-label
+	FilesystemLabel string `json:"filesystem-label"`
+	// PartitionLabel is the label of the partition for GPT disks, i.e.
+	// /dev/disk/by-partlabel
+	PartitionLabel string `json:"partition-label"`
+	// FilesystemUUID is the UUID of the filesystem on the partition, i.e.
+	// /dev/disk/by-uuid
+	FilesystemUUID string `json:"filesystem-uuid"`
+	// ID is the partition ID of the partition, i.e. /dev/disk/by-id which is
+	// typically only present for DOS disks.
+	ID string `json:"id"`
+	// Offset is the offset of the structure
+	Offset quantity.Offset `json:"offset"`
+	// Size is the size of the structure
+	Size quantity.Size `json:"size"`
+}
+
+// DiskVolumeDeviceTraits saves the mapping of volume names to volume / device
```

Comment:
missing Save... ?  should Volume be plural here given the map?

Reply:
fixed